File: katas/csharp/OOP/encapsulation/encapsulation-bad.cs
using System;
using System.Collections.Generic;

// Violación de Encapsulamiento: Datos expuestos sin protección
// ❌ Problema: Cualquiera puede modificar el estado sin validación

public class BankAccount
{
  // ❌ Propiedades públicas permiten acceso directo sin control
  public decimal Balance;
  public string AccountNumber;
  public bool IsActive;
  public List<string> TransactionHistory;

  public BankAccount(string accountNumber, decimal initialBalance)
  {
    AccountNumber = accountNumber;
    Balance = initialBalance;
    IsActive = true;
    TransactionHistory = new List<string>();
  }

  // ❌ Método sin validaciones
  public void Deposit(decimal amount)
  {
    Balance += amount;
    TransactionHistory.Add($"Depósito: {amount}");
  }

  // ❌ Método sin validaciones
  public void Withdraw(decimal amount)
  {
    Balance -= amount;
    TransactionHistory.Add($"Retiro: {amount}");
  }
}

// ❌ Sin encapsulamiento, cualquiera puede romper las reglas
class Program
{
  static void Main()
  {
    Console.WriteLine("=== Violación de Encapsulamiento ===");

    var account = new BankAccount("001", 1000);
    Console.WriteLine($"Balance inicial: ${account.Balance}");

    // ❌ Modificación directa sin validación
    account.Balance = -5000; // ¡Balance negativo!
    Console.WriteLine($"Balance después de modificación directa: ${account.Balance}");

    // ❌ Cambiar número de cuenta directamente
    account.AccountNumber = "999"; // ¡No debería ser posible!
    Console.WriteLine($"Número de cuenta modificado: {account.AccountNumber}");

    // ❌ Desactivar cuenta sin proceso adecuado
    account.IsActive = false; // Bypass de reglas de negocio
    Console.WriteLine($"Cuenta activa: {account.IsActive}");

    // ❌ Modificar historial directamente
    account.TransactionHistory = new List<string>(); // ¡Borrar evidencia!
    Console.WriteLine($"Historial manipulado: {account.TransactionHistory.Count} transacciones");

    // ❌ Retiro sin validación de fondos suficientes
    account.Balance = 100;
    account.Withdraw(1000); // ¡Retiro mayor al balance!
    Console.WriteLine($"Balance después de retiro excesivo: ${account.Balance}");
  }
}

File: katas/typescript/src/OOP/encapsulation/encapsulation-bad.ts
// Violación de Encapsulamiento: Datos expuestos sin protección
// ❌ Problema: Cualquiera puede modificar el estado sin validación

class BankAccount {
  // ❌ Propiedades públicas permiten acceso directo sin control
  public balance: number;
  public accountNumber: string;
  public isActive: boolean;
  public transactionHistory: string[];

  constructor(accountNumber: string, initialBalance: number) {
    this.accountNumber = accountNumber;
    this.balance = initialBalance;
    this.isActive = true;
    this.transactionHistory = [];
  }

  // ❌ Método sin validaciones
  public deposit(amount: number): void {
    this.balance += amount;
    this.transactionHistory.push(`Depósito: ${amount}`);
  }

  // ❌ Método sin validaciones
  public withdraw(amount: number): void {
    this.balance -= amount;
    this.transactionHistory.push(`Retiro: ${amount}`);
  }
}

// ❌ Sin encapsulamiento, cualquiera puede romper las reglas
console.log("=== Violación de Encapsulamiento ===");

const account = new BankAccount("001", 1000);
console.log(`Balance inicial: $${account.balance}`);

// ❌ Modificación directa sin validación
account.balance = -5000; // ¡Balance negativo!
console.log(`Balance después de modificación directa: $${account.balance}`);

// ❌ Cambiar número de cuenta directamente
account.accountNumber = "999"; // ¡No debería ser posible!
console.log(`Número de cuenta modificado: ${account.accountNumber}`);

// ❌ Desactivar cuenta sin proceso adecuado
account.isActive = false; // Bypass de reglas de negocio
console.log(`Cuenta activa: ${account.isActive}`);

// ❌ Modificar historial directamente
account.transactionHistory = []; // ¡Borrar evidencia!
console.log(`Historial manipulado: ${account.transactionHistory.length} transacciones`);

// ❌ Retiro sin validación de fondos suficientes
account.balance = 100;
account.withdraw(1000); // ¡Retiro mayor al balance!
console.log(`Balance después de retiro excesivo: $${account.balance}`);

export { BankAccount };
